### PR TITLE
DLS-13115: Add new HIP eligibility endpoint

### PR DESCRIPTION
**Pull request by ShivangiShah42**
Note: Exisiting logic of the Eligibility check is reused and some duplication from the DES Conrtoller exists, which can be cleaned up after the switch over to the HIP API

Also the basic postman collection attached to the ticket

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+          case None =>
+            request.body.asJson match {
+              case None =>
+                logger.warn("[EligibilityCheckHipController] - no JSON in body")
+                BadRequest
+
+              case Some(json) =>
+                json
+                  .validate[EligibilityCheckHipRequest]
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala:72]: Validation here should also check for the fixed values it has support for. Otherwise invalid request might also get processed for a success response.

File: test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Base64
+
+class EligibilityCheckHipControllerSpec extends TestSupport with AkkaMaterializerSpec {
+
+
+  val hipCredentials = "hip-client-id:hip-secret"
+
+
+
+  val hipHeaders: String = "Basic " + Base64.getEncoder.encodeToString(hipCredentials.getBytes("UTF-8"))
```
Comment (mahesh-hmrc) [inline test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala]: Could we get this from AppConfig which also maintains this and reads from application.conf, than hardcoding here again.

File: app/uk/gov/hmrc/helptosavestub/controllers/HIPController.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.google.inject.Inject
+import play.api.mvc.*
+import uk.gov.hmrc.helptosavestub.config.AppConfig
+import uk.gov.hmrc.helptosavestub.util.Logging
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import scala.concurrent.Future
+
+class HIPController  @Inject()(cc: ControllerComponents, appConfig: AppConfig)
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/HIPController.scala]: This is very much same as DESController, we could reduce duplication by making use of existing DESController by renaming it and passing in the appConfig header key to authorised action function.  So, we can rename DESController to something that reflects any downstream system. 

Discussion of keeping some duplication was mainly for the rule configurations as part of the eligibility controller.

File: app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        withHipCorrelationID(response, request.headers.get("correlationId").getOrElse(UUID.randomUUID().toString))
+      }
+    }
+
+  private def getResponse(
+                           nino: String,
+                           universalCreditClaimant: Option[String],
+                           withinThreshold: Option[String]): Result = {
+    val upperCaseNINO = nino.toUpperCase()
+    // Private BETA:
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala]: we can drop "private beta" comments. happy to retain the other ones describing the intent of codes.

File: app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    // Start NINO with EL08 to specify an eligible applicant in receipt of WTC and UC (with reason code 8)
+    if (upperCaseNINO.startsWith("EL")) {
+      Ok(toHipResult(eligibleResult(getReasonCodeFromNino(nino))).toJson)
+    } else if (upperCaseNINO.startsWith("AC")) {
+      Ok(toHipResult(alreadyHasAccountResult).toJson)
+    } else if (upperCaseNINO.startsWith("DS01")) {
+      Ok(manualRejectionResult.toJson)
+    } // Start NINO with TM02 to force eligibility check to time out
+    else if (upperCaseNINO.startsWith("TM02")) {
+      Thread.sleep(90000)
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala:102]: do we know the reason for this Thread.sleep? Anyway we can drop this ?

File: test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+
+  def fakeRequest (ucClaimant: Option[Boolean], withinThreshold: Option[Boolean]): FakeRequest[AnyContentAsJson] = {
+    val request = EligibilityCheckHipRequest("NINO Not Found", "Current Award includes a WTC entitlement", 0, 0, ucClaimant, withinThreshold)
```
Comment (mahesh-hmrc) [inline test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala]: Request details don't align with the defaults mentioned in the JIRA (that are aligned with the spec). Also, maybe we could test negative scenario when any other defaults are passed in. And, this validation could be part of the overall request validation we do (once the validation comment raised at controller level is fixed).

File: conf/prod.routes [hunk 7]
```diff
@@ -10,6 +10,10 @@ PUT        /help-to-save/accounts/:nino               uk.gov.hmrc.helptosavestub
 GET        /help-to-save/eligibility-check/:nino      uk.gov.hmrc.helptosavestub.controllers.EligibilityCheckController.eligibilityCheck(nino, universalCreditClaimant: Option[String], withinThreshold: Option[String])
 GET        /universal-credits/threshold-amount        uk.gov.hmrc.helptosavestub.controllers.DESThresholdController.getThresholdAmount()
 
+# HIP Elligibility check
+POST       /help-to-save/:identifier/account         uk.gov.hmrc.helptosavestub.controllers.EligibilityCheckHipController.eligibilityCheck(identifier: String)
```
Comment (mahesh-hmrc) [inline conf/prod.routes:14]: Can we keep README updated with inclusion of this new endpoint

File: app/uk/gov/hmrc/helptosavestub/controllers/HIPController.scala (hunk 3 above)
Comment (ShivangiShah42) [reply app/uk/gov/hmrc/helptosavestub/controllers/HIPController.scala]: Hi Mahesh
IF also have separate IFcontroller for auth, but happy to combine DES/HIP togethter considering similar API.

Comment (mahesh-hmrc) [reply app/uk/gov/hmrc/helptosavestub/controllers/HIPController.scala]: Yeah, lets tidy up pls as we have a chance here. Hopefully by the time we finish these HIP migrations we can get code base into a decent shape.

**Review by mahesh-hmrc — DISMISSED** [review]

File: app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+                           universalCreditClaimant: Option[String],
+                           withinThreshold: Option[String]): Result = {
+    val upperCaseNINO = nino.toUpperCase()
+    if (upperCaseNINO.startsWith("EL")) {
+      Ok(toHipResult(eligibleResult(getReasonCodeFromNino(nino))).toJson)
+    } else if (upperCaseNINO.startsWith("AC")) {
+      Ok(toHipResult(alreadyHasAccountResult).toJson)
+    } else if (upperCaseNINO.startsWith("DS01")) {
+      Ok(manualRejectionResult.toJson)
+    } else if (upperCaseNINO.startsWith("TM02")) {
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala:101]: Any particular reason TM02 doesn't anymore mimic timeout scenario? DES path uses this specifically for that.

File: test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+      s"  \"workingTaxCreditEntitlement\" : \"$entitlement\"," +
+      "  \"workingTaxCreditTaperedHouseholdAward\" : 0," +
+      "  \"childTaxCreditTaperedHouseholdAward\" : 0" +
+      "}"
+    FakeRequest("POST", "/")
+      .withHeaders("Authorization" -> hipHeaders)
+      .withHeaders("correlationId" -> correlationId)
+      .withHeaders("gov-uk-originator-id" -> goUKOriginatorId)
+
+      .withJsonBody(Json.toJson(request))
```
Comment (mahesh-hmrc) [inline test/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipControllerSpec.scala]: We are double encoding request here, which means tests using this might be false positives. Pls check

File: app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+  )
+
+  private val resultMap: Map[Int, String] = Map(
+    1 -> "CUSTOMER ELIGIBLE FOR HTS ACCOUNT",
+    2 -> "CUSTOMER INELIGIBLE FOR HTS ACCOUNT"
+  )
+
+  private def toHipResult(r: EligibilityCheckResult): EligibilityCheckHipResult = {
+    EligibilityCheckHipResult(
+      eligibilityResult = if(r.resultCode == 1) resultMap(1) else resultMap(2),
```
Comment (mahesh-hmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala]: Could we not convert every non-eligible DES result code into a valid HIP ineligible response? Can we make sure there's handlers for WP99 profiles, as they are cases for unknown eligibility state. Which is different to not-eligible.

Comment (ShivangiShah42) [reply app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala]: Hi Mahesh
EligibilityResult contains only two possible values 
Enum: "CUSTOMER ELIGIBLE FOR HTS ACCOUNT" "CUSTOMER INELIGIBLE FOR HTS ACCOUNT"

Comment (mahesh-hmrc) [reply app/uk/gov/hmrc/helptosavestub/controllers/EligibilityCheckHipController.scala]: Unknown state, if for any reason we get from backend, should be treated as an error state than mark that as not-eligible.

**Review by mahesh-hmrc — APPROVED** [review]
